middle_end/flambda2/to_cmm/to_cmm_expr.ml

```diff
@@ -265,15 +267,39 @@ let translate_apply0 ~dbg_with_inlined:dbg env res apply =
     in
     let effects = To_cmm_effects.transl_c_call_effects effects in
     let coeffects = To_cmm_effects.transl_c_call_coeffects coeffects in
-    ( wrap dbg
-        (C.extcall ~dbg ~alloc:needs_caml_c_call ~is_c_builtin ~effects
-           ~coeffects ~returns ~ty_args callee
-           (C.Extended_machtype.to_machtype return_ty)
-           args),
-      free_vars,
-      env,
-      res,
-      Ece.all )
+    let extcall =
+      C.extcall ~dbg ~alloc:needs_caml_c_call ~is_c_builtin ~effects ~coeffects
+        ~returns ~ty_args callee return_ty args
+    in
+    let wrap return_values =
+      let kinds = Flambda_arity.unarized_components return_arity in
+      assert (List.compare_length_with kinds (Array.length component_tys) = 0);
+      match kinds with
+      | [] -> return_values
+      | [kind] -> maybe_sign_extend kind dbg return_values
+      | [_; _] as kinds ->
+        (* CR xclerc: we currently support only pairs as unboxed return values. *)
+        let get_unarized_return_value exp n =
+          C.tuple_field exp ~component_tys n dbg
+        in
+        C.make_tuple
+          (List.mapi
+             (fun i kind ->
+               maybe_sign_extend kind dbg
+                 (get_unarized_return_value return_values i))
+             kinds)
+      | _ ->
+        Misc.fatal_errorf
+          "C functions are currently limited to a single return value or a \
+           pair of return values"
```

Comment:
could you try and split off (most of) the code for treating unboxed returns in a separate function ? The code for the `C_call` case is starting to be quite long..

Reply:
I've moved the whole thing, indeed it is getting lengthy.